Fixes #21026 - Clear cached notifications properly

bastilian opened the pull request:
Using `delete_all` for has_many does not trigger callbacks
on the associated objects, `dependent: :destroy` does.

Clearing the cache needs to happen anytime a notification
gets created, modified and destroyed, `after_commit` will
be triggered on all o these occasions.

mmoll commented:
please rebase :)

iNecas commented:
It looks good to me, but would rather have @ohadlevy or @tbrisker confirmation this is how it's supposed to work

iNecas commented:
test failures unrealted

mmoll commented on app/models/notification.rb:
```diff
@@ -15,8 +15,8 @@ class Notification < ApplicationRecord
-  has_many :notification_recipients, :dependent => :delete_all
-  has_many :recipients, :class_name => 'User', :through => :notification_recipients, :source => :user
+  has_many :notification_recipients, :dependent => :destroy
+  has_many :recipients, :class_name => User, :through => :notification_recipients, :source => :user
```
this should be `:class_name => 'User'`

bastilian committed:
Fixes #21026 - Clear cached notifications properly

Using `delete_all` for has_many does not trigger callbacks
on the associated objects, `dependent: :destroy` does.

Clearing the cache needs to happen anytime a notification
gets created, modified and destroyed, `after_commit`
will be triggered on all o these occasions.

tbrisker commented:
Thanks @bastilian !
